File: python/sieve_of_eratosthenes.py
def get_primes_up_to(n : int) -> list[int]:
    primes = []

    if n < 2:
        return primes  # No primes less than 2

    # Create a boolean list to track prime numbers
    is_prime = [True] * (n + 1)
    is_prime[0] = is_prime[1] = False  # 0 and 1 are not prime

    # Sieve of Eratosthenes
    for i in range(2, int(n ** 0.5) + 1):
        if is_prime[i]:
            for j in range(i * i, n + 1, i):
                is_prime[j] = False  # Mark multiples of i as non-prime

    # Collect all prime numbers into a list
    for i in range(2, n + 1):
        if is_prime[i]:
            primes.append(i)

    return primes

File: rust/src/sieve_of_eratosthenes.rs
pub fn get_primes_up_to(n: usize) -> Vec<u32> {

    if n < 2 {
        panic!("In the function get_primes_up_to(n) n shouldn't be less than 2. n = {}", n);
    }

    // Create a boolean vector to track prime numbers
    let mut is_prime: Vec<bool>  = vec![true; n + 1];
    is_prime[0] = false; // 0 is not prime
    is_prime[1] = false; // 1 is not prime

    // Sieve of Eratosthenes
    for i in 2..=n.isqrt() {
        if is_prime[i] {
            for j in (i * i..=n).step_by(i) {
                is_prime[j] = false; // Mark multiples of i as non-prime
            }
        }
    }

    is_prime.iter()
            .enumerate()
            .filter(|(_, is_prime)| **is_prime )
            .map(|(index, _)| index as u32)
            .collect()
}
    